### PR TITLE
Merge preload and present choice set operations

### LINKED ISSUE #473: Wrong ChoiceSet listeners being called when queueing PresentChoiceSetOperations
### Bug Report
If multiple PresentChoiceSetOperations are queued with the ChoiceSetManager, then the listeners from later ChoiceSets can end up being called on earlier ChoiceSets. This is likely because the operation class is using a reference to the manager's current pending ChoiceSet instead of cloning the data.

##### Reproduction Steps
1. Present a ChoiceSet with the manager and while it's being presented, queue another ChoiceSet to be presented. Provide an `onChoiceSelected` listener for both.
2. In the HMI, select one of the available choices from the first ChoiceSet.

##### Expected Behavior
The listener for the first ChoiceSet is called.

##### Observed Behavior
The listener for the second ChoiceSet is called.

### PR DESCRIPTION
**Pull request by renonick87**
Fixes #473, #474, #478

### Risk
This PR makes **minor** API changes.

### Testing Plan
- [X] I have verified that I have not introduced new warnings in this PR (or explain why below)
- [X] I have verified that this PR passes lint validation
- [X] I have run the unit tests with this PR
- [X] I have tested this PR against Core and verified behavior (if applicable, if not applicable, explain why below).

### Summary
This PR fixes several issues that were present in the ChoiceSetManager, including:
- Merging the preload and present Choices operations.
- Taking preloaded cells into account when creating unique text for new cells.
- Updating pending tasks with the most recent list of preloaded choices.
- Moved the choice uniqueness logic to the operation class.
- The wrong ChoiceSet listeners were called if multiple operations were queued.
- Present followed by delete could either delete some Choices after they were preloaded but before they were presented, or could attempt to delete choices that had not been loaded.

### CLA
- [X] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: lib/js/src/manager/screen/choiceset/_PreloadPresentChoicesOperation.js [hunk 1]
```diff
@@ -0,0 +1,824 @@
...
+        // internal usage
+        this._isRunning = false;
+    }
+
+    /**
+     * The method that causes the task to run.
+     * @param {_Task} task - The task instance
+     * @returns {Promise} - Does not resolve to any value
+     */
+    async onExecute (task) {
```
Comment (renonick87) [inline lib/js/src/manager/screen/choiceset/_PreloadPresentChoicesOperation.js:113]: The onExecute and _startPresent methods need to be updated for alignment. These methods should be checking the return values from methods such as preloadCells or updateKeyboardProperties and canceling the present if an error occurs along the way.

**Review by crokita — CHANGES_REQUESTED** [review]
See feedback

File: lib/js/src/manager/screen/choiceset/_ChoiceSetManagerBase.js [hunk 2]
```diff
@@ -138,53 +131,23 @@ class _ChoiceSetManagerBase extends _SubManagerBase {
...
-            }
-        });
+        const preloadChoicesOperation = new _PreloadPresentChoicesOperation(this._lifecycleManager, this._fileManager, this._displayName,
+            this._defaultMainWindowCapability, this._isVrOptional, choices,
+            this._preloadedChoices, null, null, null, null, null, null, (loadedCells, success) => {
+                if (!success) {
+                    console.error('ChoiceSetManager: There was an error preloading choice cells');
+                    return;
+                }
+                this._preloadedChoices = loadedCells;
```
[the remark this answers is not in the document]
Comment (renonick87) [reply lib/js/src/manager/screen/choiceset/_ChoiceSetManagerBase.js]: This wouldn't quite work since it's happening within the operation's completion listener. Instead I enclosed the operation in a promise similar to how it's done with deleteChoices.

**Review by crokita — CHANGES_REQUESTED** [review]
See feedback

File: lib/js/src/manager/screen/choiceset/_PreloadPresentChoicesOperation.js [hunk 3]
```diff
@@ -0,0 +1,840 @@
...
+    // HELPER METHODS
+    /**
+     * Creates a PerformInteraction RPC
+     * @returns {PerformInteraction} - The PerformInteraction RPC
+     */
+    _getPerformInteraction () {
+        const pi = new PerformInteraction()
+            .setInitialText(this._choiceSet.getTitle())
+            .setInteractionMode(this._presentationMode)
+            .setInteractionChoiceSetIDList(this._choiceSet.getChoices().map(choiceCell => choiceCell._getChoiceId()))
```
[the remark this answers is not in the document]
Comment (renonick87) [reply lib/js/src/manager/screen/choiceset/_PreloadPresentChoicesOperation.js:641]: In regards to the second comment, the IDs of the choices should be getting updated in `_updateChoiceSetChoicesId` which uses the list of loaded cells to update the IDs on the choices being presented. It seems that this method wasn't taking unique text into account, which could cause issues. I'll do additional digging to see if there are other issues here.

[the remark this answers is not in the document]
Comment (renonick87) [reply lib/js/src/manager/screen/choiceset/_PreloadPresentChoicesOperation.js:627]: I fixed some of the logic before the operation is created as well as before the cells are preloaded. This issue is no longer appearing.

**Review by crokita — COMMENTED** [review]
See question

File: lib/js/src/manager/screen/choiceset/_ChoiceSetManagerBase.js [hunk 4]
```diff
@@ -131,62 +124,39 @@ class _ChoiceSetManagerBase extends _SubManagerBase {
...
-                        this._removeChoicesFromChoices(choicesToUpload, this._pendingPreloadChoices);
+            const preloadChoicesOperation = new _PreloadPresentChoicesOperation(this._lifecycleManager, this._fileManager, this._displayName,
+                this._defaultMainWindowCapability, this._isVrOptional, choices,
+                this._preloadedChoices, null, null, null, null, null, null, (loadedCells, success) => {
+                    if (!success) {
+                        console.error('ChoiceSetManager: There was an error preloading choice cells');
+                        return resolve(false);
                     }
-                    resolve(success);
+                    this._preloadedChoices = loadedCells.map(cell => cell.clone());
```
[the remark this answers is not in the document]
Comment (renonick87) [reply lib/js/src/manager/screen/choiceset/_ChoiceSetManagerBase.js]: The cloning does need to happen in the other operations, however doing this at the end of the operation caused issues since an operation could be started and the cells could be updated while the task is running (in some cases this would happen after the cells had been preloaded but before being passed back to the manager). I've moved the cloning to happen during the `_updateCells...` methods which are called at the start of the operations.

**Comment by renonick87** [comment]
I also fixed an issue with the DeleteChoices operation where the task would hang if there were no valid choices to be deleted.

**Review by crokita — APPROVED** [review]

File: lib/js/src/manager/screen/choiceset/_PreloadPresentChoicesOperation.js [hunk 5]
```diff
@@ -0,0 +1,920 @@
...
+                cellIdLists[duplicateIndex].push(cellUniqueId);
+            } else {
+                cells.push(cellKey);
+                cellIdLists.push([cellUniqueId]);
+            }
+        });
+
+        for (const cellIdList of cellIdLists) {
+            cellIdList.sort((id1, id2) => {
+                return id1 - id2;
```
Comment (crokita) [inline lib/js/src/manager/screen/choiceset/_PreloadPresentChoicesOperation.js:317]: I cannot believe this works. Classic JavaScript

**Review by crokita — APPROVED** [review]

**Comment by renonick87** [comment]
The logic for `_PreloadPresentChoicesOperation`'s `_assignIdsToCells` method has been updated to account for Ids used by previous cells that haven't been loaded yet, as well as better handling of edge cases where a developer may have reached the maximum choice ID available. Unit tests have been added to cover these cases.
There was also a bug fixed with how the operation checked whether Core supported choice uniqueness. The value returned was the inverse of what was actually desired.

**Review by crokita — CHANGES_REQUESTED** [review]
See minor feedback

**Comment by renonick87** [comment]
Due to a Generic HMI bug, if a user selected a Choice with ID=0 then the PerformInteractionResponse would be received without a corresponding choice ID property. Since we don't know if there are any production HMIs that also have this bug, and since the minimum ID prior to this refactor was already set to 1, we've adjusted the minimum choice ID back to 1.

**Review by crokita — APPROVED** [review]

**Review by crokita — APPROVED** [review]

**Review by crokita — APPROVED** [review]
